File: data_structures/linked_list/doubly_linked_list.py
class Node:

	def __init__(self, data=None, next=None):
		## data of the node
		self.data = data
		## next pointer
		self.next = next
	
	def get_data(self):
		return self.data

	def get_next(self):
		return self.next
	
	def set_next(self, new_next):
		self.next = new_next

class LinkedList:

	def __init__(self):
		## initializing the head with None
		self.head = None

	def insert(self, data):
		## check whether the head is empty or not
		new_node = Node(data)

		if self.head:
			## getting the last node
			last_node = self.head
			while last_node.next != None:
				last_node = last_node.next

			## assigning the new node to the next pointer of last node
			last_node.next = new_node

		else:
			## head is empty
			## assigning the node to head
			self.head = new_node
	
	def size(self):
		current = self.head
		cnt = 0

		while current:
			cnt += 1
			current = current.get_next()
		
		print(cnt)

	def search(self, data):
		current = self.head
		found = False

		while current and found is False:
			if current.get_data() == data:
				found = True
			else:
				current = current.get_next()
		
		if current is None:
			raise ValueError("Data not in list")
		
		print(current)

	def delete(self, data):
		current = self.head
		previous = None
		found = False

		while current and found is False:
			if current.get_data() == data:
				found = True
			else:
				previous = current
				current = current.get_next()
		
		if current is None:
			raise ValueError("Data not in list")

		if previous is None:
			self.head = current.get_next()
		else:
			previous.set_next(current.get_next())

	def display(self):
		## variable for iteration
		temp_node = self.head

		## iterating until we reach the end of the linked list
		while temp_node != None:
			## printing the node data
			print(temp_node.data, end='->')

			## moving to the next node
			temp_node = temp_node.next

		print('Null')


if __name__ == '__main__':
	## instantiating the linked list
	linked_list = LinkedList()

	## inserting the data into the linked list
	linked_list.insert(1)
	linked_list.insert(2)
	linked_list.insert(3)
	linked_list.insert(4)
	linked_list.insert(5)
	linked_list.insert(6)
	linked_list.insert(7)

	## printing the linked list
	linked_list.display()

	linked_list.size()

	linked_list.search(6)

	linked_list.delete(6)

	linked_list.display()

File: data-structures/linked-list/DoublyLinkedList.java
import java.util.List;
import java.util.ArrayList;

public class SinglyLinkedList<T> {
	public static class Node<T> {
		private T data;
		private Node<T> next;
		private Node<T> prev;

		public Node() {
			this.data = null;
			this.next = null;
			this.prev = null;
		}

		public Node(T data) {
			this.data = data;
		}
	}

	private Node<T> head;
	private int size;

	public SinglyLinkedList() {
		this.head = null;
	}

	public void insert(T data) {
		Node<T> insert = new Node<>(data);
		if(this.head == null) {
			this.head = insert;
			size++;
			return;
		}
		Node<T> curr = this.head;
		while(curr.next != null) {
			curr = curr.next;
		}
		curr.next = insert;
		size++;
	}

	public void delete(T data) {
		if(this.head == null) return;
		Node<T> curr = this.head;
		Node<T> prev = this.head;
		while(curr.next != null) {
			if(curr.data == data) {
				prev.next = curr.next;
				size--;
				break;
			}
			prev = curr;
			curr = curr.next;
		}
	}

	public int search(T data) {
		Node<T> curr = this.head;
		int cnt = 0;
		while(curr.next != null) {
			if(curr.data == data) return cnt;
			cnt++;
			curr = curr.next;
		}
		return -1;
	}

	public T get(int index) {
		Node<T> curr = this.head;
		int cnt = 0;
		while(curr.next != null && cnt != index && cnt < size) {
			cnt++;
			curr = curr.next;
		}
		if(cnt == index) return curr.data;
		else return null;
	}

	public int size() {
		return this.size;
	}

	public String toString() {
		List<T> list = new ArrayList<>();
		Node<T> curr = this.head;
		while(curr != null) {
			list.add(curr.data);
			curr = curr.next;
		}
		return list.toString();
	}

	public static void main(String[] args) {
		SinglyLinkedList<Integer> ll = new SinglyLinkedList<>();
		System.out.println(ll.toString());
		ll.insert(4);
		System.out.println(ll.toString());
		ll.insert(2);
		ll.insert(18);
		ll.insert(43);
		ll.insert(993);
		ll.insert(34);
		System.out.println(ll.toString());
		System.out.println(ll.search(18));
		System.out.println(ll.get(2));
		System.out.println(ll.size());
	}
}